profiling/src/profile/mod.rs

```diff
@@ -491,6 +531,84 @@ impl Profile {
             .add_endpoint_count(endpoint.into_owned(), value);
     }
 
+    pub fn add_upscaling_rules(
+        &mut self,
+        offset_values: &[usize],
+        label_name: &str,
+        label_value: &str,
+        upscaling_info: UpscalingInfo,
+    ) -> anyhow::Result<()> {
+        let label_name_id = self.intern(label_name);
+        let label_value_id = self.intern(label_value);
+
+        let mut new_values_offset = offset_values.to_vec();
+        new_values_offset.sort_unstable();
+
+        fn is_overlapping(v1: &[usize], v2: &[usize]) -> bool {
+            v1.iter().any(|x| v2.contains(x))
+        }
+
+        anyhow::ensure!(
+            self.upscaling_rules
+                .iter()
+                .all(|(_, rules)| rules.iter().all(|rule| !is_overlapping(
+                    &rule.values_offset,
+                    &new_values_offset
+                ) || rule.label_name != label_name_id
+                    || rule.label_value != label_value_id)),
+            "Duplicated upscaling rule"
+        );
```

Comment:
I wonder: can this get a bit expensive? I'm assuming that for instance in the exception profiling case we may call this method hundreds or thousands of times (for every unique exception name in the app).

(Would it be preferable to perhaps only this validation inline with serialization? Or perhaps do it in a single pass after all rules were added?)

Reply:
This can be a bit expensive, I changed a bit the implementation, so for `by-label` rules, this should be more efficient.
For the exception profiling, you will call this as much as unique exception name you have. For .NET application, we do not throw a lot of unique exception name, but maybe other can.

> Would it be preferable to perhaps only this validation inline with serialization? Or perhaps do it in a single pass after all rules were added?

Catching rule collision when adding can provide more information on what failed (again the new implementation tries to achieve that)

Reply:
(Not blocking for this PR!)

The newer version indeed is better but... one thing that I'm not a big fan of is that for instance the checks penalize adding by-value first, and then by-label later -- because in this case we check every with-label against all the by-value rules; it's much better to add all by-label first (check checks) and then by-value later.

But... the API doesn't really push clients to do that. So clients can get a more expensive behavior if they haven't noticed how the validation code works.

And that's.... a bit of a sharp edge. I guess we can document it, but it would be nice if either the checks either were independent of API calling order, of it the API forced clients to use it correctly -- by perhaps not allowing the addition of by-value rules after by-label rules have been added.